File: sx126x.py
import RPi.GPIO as GPIO
import time
import serial
import os

class sx126x():

    SERIAL_PORT_RATE = {'1200' : 0b00000000, '2400' : 0b00100000, '4800' : 0b01000000, '9600' : 0b01100000, '19200' : 0b10000000, '38400' : 0b10100000, '57600' : 0b11000000, '115200' : 0b11100000 }
    SERIAL_PARITY_BIT = {'8N1' : 0b00000000, '8O1' : 0b00001000, '8E1' : 0b00010000 } #serial.PARITY_NONE - 8N1, serial.PARITY_ODD - 8O1, serial.PARITY_EVEN - 8E1    
    AIR_DATA_RATE = {'0.3' : 0b00000000, '1.2' : 0b00000001, '2.4' : 0b00000010, '4.8' : 0b00000011, '9.6' : 0b00000100, '19.2' : 0b00000101, '38.4' : 0b00000110, '62.5' : 0b00000111 }
    SUB_PACKET_SIZE = {'240' : 0b00000000, '128' : 0b01000000, '64' : 0b10000000, '32' : 0b11000000 }
    CHANNEL_NOISE = {'off' : 0b00000000, 'on' : 0b00100000 }
    TX_POWER = {'22' : 0b00000000, '17' : 0b00000001, '13' : 0b00000010, '10' : 0b00000011 }
    ENABLE_RSSI = {'off' : 0b00000000, 'on' : 0b10000000 }
    TRANSMISSION_MODE = {'fixedPoint' : 0b01000000, 'transparent' : 0b000000000 }
    ENABLE_REPEATER = {'off' : 0b00000000, 'on' : 0b00100000 }
    ENABLE_LBT = {'off' : 0b00000000, 'on' : 0b00010000 }
    WOR_CONTROL = {'receiver' : 0b00001000, 'transmitter' : 0b00000000 }
    WOR_CYCLE = {'500' : 0b00000000, '1000' : 0b00000001, '1500' : 0b00000010, '2000' : 0b00000011, '2500' : 0b00000100, '3000' : 0b00000101, '3500' : 0b00000110, '4000' : 0b00000111 }


    def __init__(self, address = 100, network=0, channel=18, txPower='22', enableRSSI=False,
                    airDataRate='2.4', repeater='none', packetSize='128', debug=False):

        self.debug = debug
        self.logicalAddress = address

        #set default parameters
        self.port = '/dev/ttyS0'
        self.serialPortRate = '9600'
        self.timeout = 1
        self.serialParityBit = serial.PARITY_NONE
        self.loraParityBit = self.convertSerialParity(self.serialParityBit)
        
        self.channel = int(channel) #0 - 80
        self.crypt = 0 #0 - 65535
        self.network = int(network) #0 - 255

        self.airDataRate =airDataRate
        self.subPacketSize ='128'
        self.channelNoise ='off'
        self.txPower = str(txPower)
        self.enableRSSI = 'on' if enableRSSI else 'off'

        if repeater == "server":
            self.transmissionMode = 'fixedPoint'
            self.address = address
            self.enableRepeater = 'on'
        elif repeater == "client":
            self.transmissionMode = 'fixedPoint'
            self.address = address
            self.enableRepeater = 'off'
        else: #none
            self.transmissionMode = 'transparent'
            self.address = 65535
            self.enableRepeater = 'off'

        self.enableLBT = 'off'
        self.WORcontrol = 'transmitter'
        self.WORcycle = '2000'
        
        
        self.M0 = 22
        self.M1 = 27

        GPIO.setmode(GPIO.BCM)
        GPIO.setwarnings(False)
        GPIO.setup(self.M0, GPIO.OUT)
        GPIO.setup(self.M1, GPIO.OUT)

        #self.gpio_mode("let's rock")
        self.writeConfig()


    def show_config(self):
        print(f'Channel {self.channel}, address {self.logicalAddress}, network {self.network}')
        print(f'mode {self.transmissionMode}, real address {self.address}, repeater {self.enableRepeater}')


    def gpio_mode(self, mode):
        if mode == 'conf':
            GPIO.output(self.M0, False)
            GPIO.output(self.M1, True)
        elif mode == 'wor':
            GPIO.output(self.M0, True)
            GPIO.output(self.M1, False)
        elif mode == 'sleep':
            GPIO.output(self.M0, True)
            GPIO.output(self.M1, True)
        else:
            GPIO.output(self.M0, False)
            GPIO.output(self.M1, False)

        time.sleep(.3)


    def openSerial(self):
        ser = serial.Serial(port=self.port, baudrate=self.serialPortRate, \
                    parity=self.serialParityBit, stopbits=serial.STOPBITS_ONE,
                    bytesize=serial.EIGHTBITS, timeout=self.timeout)

        return ser


    def sendraw(self, data):
        ser = self.openSerial()

        ser.write(data)
        time.sleep(.1)
        #ret = ser.readlines()
        ser.close()


    def sendmsg(self, data):
        if self.transmissionMode == 'fixedPoint':
            data = self.logicalAddress.to_bytes(2, 'big') + self.network.to_bytes(1, 'big') + data.encode()
        else:
            data = self.logicalAddress.to_bytes(2, 'big') + data.encode()

        if self.debug:
            print('[+] sending :', data)
        
        self.sendraw(data)


    def receive(self):
        ser = self.openSerial()

        data = ser.read_until(expected='')

        if not data:
            data = None
        else:
            if self.debug:
                print('[+] received',len(data),'data :', data)

        ser.close()
        return data


    def convertSerialParity(self,parity):
        if parity == serial.PARITY_NONE:
            return "8N1"
        elif parity == serial.PARITY_ODD:
            return "8O1"
        elif parity == serial.PARITY_EVEN:
            return "8E1"
        else:
            return "8N1"


    def getRSSI(self):
        
        if self.channelNoise == 'off':
            return

        ser = self.openSerial()
        
        ser.write(b'\xC0\xC1\xC2\xC3\x00\x02')
        ret = ser.read_until()
        print(ret)
        currentNoise = ret[3]
        lastReceive = ret[4]

        ser.close()
        return currentNoise, lastReceive



    def setConfig(self, port=None, serialPortRate=None, timeout=None, serialParityBit=None,
            channel=None,crypt=None,address=None, network=None,airDataRate=None,subPacketSize=None,
            channelNoise=None,txPower=None,enableRSSI=None,transmissionMode=None,enableRepeater=None,
            enableLBT=None,WORcontrol=None,WORcycle=None):

        self.port = port or self.port
        self.serialPortRate = serialPortRate or self.serialPortRate
        self.timeout = timeout or self.timeout
        self.serialParityBit = serialParityBit or self.serialParityBit
        self.loraParityBit = self.convertSerialParity(self.serialParityBit)
        
        self.channel = channel or self.channel
        self.crypt = crypt or self.crypt
        self.address = address or self.address
        self.network = network or self.network

        self.airDataRate = airDataRate or self.airDataRate
        self.subPacketSize = subPacketSize or self.subPacketSize
        self.channelNoise = channelNoise or self.channelNoise
        self.txPower = txPower or self.txPower
        self.enableRSSI = enableRSSI or self.enableRSSI
        self.transmissionMode = transmissionMode or self.transmissionMode
        self.enableRepeater = enableRepeater or self.enableRepeater
        self.enableLBT = enableLBT or self.enableLBT
        self.WORcontrol = WORcontrol or self.WORcontrol
        self.WORcycle = WORcycle or self.WORcycle

        self.writeConfig()


    def writeConfig(self):

        #factory reset
        #C0 00 09 12 34 00 61

        RESERVE = 0b00000000

        #C0 = config, 00 = start address, 09 : length
        config = bytearray(b'\xC0\x00\x09')

        address_tmp = self.address.to_bytes(2, 'big')
        config.append(address_tmp[0])
        config.append(address_tmp[1])

        #config.append(int(addrl))

        config.append(self.network)

        config.append(int(hex(self.SERIAL_PORT_RATE[self.serialPortRate] + \
                              self.SERIAL_PARITY_BIT[self.loraParityBit] + \
                              self.AIR_DATA_RATE[self.airDataRate]), 16))


        config.append(int(hex(self.SUB_PACKET_SIZE[self.subPacketSize] + \
                              self.CHANNEL_NOISE[self.channelNoise] + \
                              RESERVE + \
                              self.TX_POWER[self.txPower]), 16))
        
        config.append(self.channel)

        config.append(int(hex(self.ENABLE_RSSI[self.enableRSSI] + 
                              self.TRANSMISSION_MODE[self.transmissionMode] + \
                              self.ENABLE_LBT[self.enableLBT] + \
                              self.WOR_CONTROL[self.WORcontrol] + \
                              self.WOR_CYCLE[self.WORcycle]), 16))

        crypt_tmp = self.crypt.to_bytes(2, 'big')
        config.append(crypt_tmp[0])
        config.append(crypt_tmp[1])
        
        #print("[+] Sending new Configuration Register", flush=True)
        #print(' '.join(['{:02X}'.format(x) for x in config]), flush=True)

        self.gpio_mode('conf')
        ser = self.openSerial()
        
        ser.write(bytes(config))
        time.sleep(.1)
        ret = ser.read_until()
        
        if ret == b'\xff\xff\xff':
            print("ERREUR CONFIG")

        ser.close()
        self.gpio_mode('')

File: Lo-Fi/sx126x/sx126x.cpp
#include "sx126x.h"


sx126x::sx126x() {
  _debug = true;
}

void sx126x::begin() {

  Serial.begin(115200);
  while (!Serial) {}

  params.serial_port_rate = 9600;
  params.serial_parity_bit = 0;
  serial_lora.begin(params.serial_port_rate, SERIAL_8N1, 18, 17);

  pinMode(M0, OUTPUT);
  pinMode(M1, OUTPUT);
  
  //very important to read here
  delay(10);
  while (serial_lora.available()) serial_lora.read();

  //read current settings
  char result[12];
  params = read_registers(result);
}

void sx126x::send(char* buffer, int count) {

  if (_debug) {
    Serial.print("[+] Sending : ");
    print_hex(buffer, count);
  }

  serial_lora.write(buffer, count);
  delay(10);
}

int sx126x::receive(char* buffer) {
  int count = 0;
  while (serial_lora.available()) {
    buffer[count++] = serial_lora.read();
    delay(1);
  }

  if (_debug) {
    //print_hex(buffer, count);
  }

  return count;
}

void sx126x::set_mode(int mode) {
  
  if (_debug) {
    Serial.print("[+] Set mode : ");
    if (mode == MODE_CONF) Serial.println("CONF");
    if (mode == MODE_TRANS) Serial.println("TRANS");
    if (mode == MODE_WOR) Serial.println("WOR");
    if (mode == MODE_SLEEP) Serial.println("SLEEP");
  }

  if (mode == MODE_CONF) {
    digitalWrite(M0, LOW);
    digitalWrite(M1, HIGH);
  }
  else if (mode == MODE_WOR) {
    digitalWrite(M0, HIGH);
    digitalWrite(M1, LOW);
  }
  else if (mode == MODE_SLEEP) {
    digitalWrite(M0, HIGH);
    digitalWrite(M1, HIGH);
  }
  else {
    digitalWrite(M0, LOW);
    digitalWrite(M1, LOW);
  }
  delay(100);
}


bool sx126x::set_address(int address)
{
  if (address < 0 || address > 65535)
    return false;

  params.addrh = address >> 8;
  params.addrl = address & 255;

  write_registers();
  return true;
}


bool sx126x::set_network(int network)
{
  if (network < 0 || network > 255)
    return false;

  params.network = network;
  write_registers();
  return true;
}

bool sx126x::set_air_data_rate(float air_data_rate)
{
  if (air_data_rate != 0.3 && air_data_rate != 1.2 && air_data_rate != 2.4 &&
    air_data_rate != 4.8 && air_data_rate != 9.6 && air_data_rate != 19.2 && 
    air_data_rate != 38.4 && air_data_rate != 62.5)
    return false;

  params.air_data_rate = (int)(air_data_rate * 10);
  write_registers();
  return true;
}

bool sx126x::set_tx_power(int tx_power)
{
  if (tx_power != 10 && tx_power != 13 && tx_power != 17 && tx_power != 22)
    return false;

  params.tx_power = tx_power;
  write_registers();
  return true;
}

bool sx126x::set_sub_packet_size(int sub_packet_size)
{
  if (sub_packet_size != 240 && sub_packet_size != 128 &&
      sub_packet_size != 64 && sub_packet_size != 32)
    return false;

  params.sub_packet_size = sub_packet_size;
  write_registers();
  return true;
}

bool sx126x::set_channel(int channel)
{
 if (channel < 0 || channel > 80)
    return false;

  params.channel = channel;
  write_registers();
  return true;
}

bool sx126x::set_channel_noise(bool channel_noise)
{
  params.channel_noise = channel_noise;
  write_registers();
  return true;
}

bool sx126x::set_enable_rssi(bool enable_rssi)
{
  params.enable_rssi = enable_rssi;
  write_registers();
  return true;
}

bool sx126x::set_crypt_key(int key)
{
 if (key < 0 || key > 65535)
    return false;

  params.crypth = key >> 8;
  params.cryptl = key & 255;
  write_registers();
  return true;
}

bool sx126x::get_channel_noise()
{
  return params.channel_noise;
}

bool sx126x::get_enable_rssi()
{
  return params.enable_rssi;
}

s_param sx126x::read_registers(char* result)
{
  char data[] = {0xC1, 0x00, 0x09};
  write_serial(data, 3, result);
  struct s_param read_params;

  read_params.addrh = result[3];
  read_params.addrl = result[4];
  read_params.network = result[5];
  read_params.channel = result[8];
  read_params.crypth = result[10];
  read_params.cryptl = result[11];

  for (int i = 0; i < 8; i++) if ((result[6] & 0b11100000) == SERIAL_PORT_RATE[i][1]) read_params.serial_port_rate = SERIAL_PORT_RATE[i][0];
  for (int i = 0; i < 3; i++) if ((result[6] & 0b00011000) == SERIAL_PARITY_BIT[i][1]) read_params.serial_parity_bit = SERIAL_PARITY_BIT[i][0];  
  for (int i = 0; i < 8; i++) if ((result[6] & 0b00000111) == AIR_DATA_RATE[i][1]) read_params.air_data_rate = AIR_DATA_RATE[i][0];
  for (int i = 0; i < 4; i++) if ((result[7] & 0b11000000) == SUB_PACKET_SIZE[i][1]) read_params.sub_packet_size = SUB_PACKET_SIZE[i][0];
  for (int i = 0; i < 2; i++) if ((result[7] & 0b00100000) == CHANNEL_NOISE[i][1]) read_params.channel_noise = (CHANNEL_NOISE[i][0] == 1);  
  for (int i = 0; i < 4; i++) if ((result[7] & 0b00000011) == TX_POWER[i][1]) read_params.tx_power = TX_POWER[i][0];
  for (int i = 0; i < 2; i++) if ((result[9] & 0b10000000) == ENABLE_RSSI[i][1]) read_params.enable_rssi = (ENABLE_RSSI[i][0] == 1);
  for (int i = 0; i < 2; i++) if ((result[9] & 0b01000000) == TRANSMISSION_MODE[i][1]) read_params.transmission_mode = (TRANSMISSION_MODE[i][0] == 1);
  for (int i = 0; i < 2; i++) if ((result[9] & 0b00100000) == ENABLE_REPEATER[i][1]) read_params.enable_repeater = (ENABLE_REPEATER[i][0] == 1);
  for (int i = 0; i < 2; i++) if ((result[9] & 0b00010000) == ENABLE_LBT[i][1]) read_params.enable_lbt = (ENABLE_LBT[i][0] == 1);
  for (int i = 0; i < 2; i++) if ((result[9] & 0b00001000) == WOR_CONTROL[i][1]) read_params.wor_control = (WOR_CONTROL[i][0] == 1);
  for (int i = 0; i < 8; i++) if ((result[9] & 0b00000111) == WOR_CYCLE[i][1]) read_params.wor_cycle = WOR_CYCLE[i][0];

  read_params.serial_port_rate = 9600;
  read_params.serial_parity_bit = 0;

  return read_params;
}


void sx126x::write_registers()
{
  char settings[12] = {0xC0, 0x00, 0x09};
  int reserve = 0;

  int serial_port_rate, serial_parity_bit, air_data_rate, sub_packet_size;
  int channel_noise, tx_power, enable_rssi, transmission_mode, enable_repeater;
  int enable_lbt, wor_control, wor_cycle;

  for (int i = 0; i < 8; i++) if (params.serial_port_rate == SERIAL_PORT_RATE[i][0]) serial_port_rate = SERIAL_PORT_RATE[i][1];
  for (int i = 0; i < 3; i++) if (params.serial_parity_bit == SERIAL_PARITY_BIT[i][0]) serial_parity_bit = SERIAL_PARITY_BIT[i][1];  
  for (int i = 0; i < 8; i++) if (params.air_data_rate == AIR_DATA_RATE[i][0]) air_data_rate = AIR_DATA_RATE[i][1];
  for (int i = 0; i < 4; i++) if (params.sub_packet_size == SUB_PACKET_SIZE[i][0]) sub_packet_size = SUB_PACKET_SIZE[i][1];
  for (int i = 0; i < 4; i++) if (params.tx_power == TX_POWER[i][0]) tx_power = TX_POWER[i][1];
  for (int i = 0; i < 8; i++) if (params.wor_cycle == WOR_CYCLE[i][0]) wor_cycle = WOR_CYCLE[i][1];  
  channel_noise = CHANNEL_NOISE[params.channel_noise][1];
  enable_rssi = ENABLE_RSSI[params.enable_rssi][1];
  transmission_mode = TRANSMISSION_MODE[params.transmission_mode][1];
  enable_repeater = ENABLE_REPEATER[params.enable_repeater][1];
  enable_lbt = ENABLE_LBT[params.enable_lbt][1];
  wor_control = WOR_CONTROL[params.wor_control][1];

  settings[3] = params.addrh;
  settings[4] = params.addrl;
  settings[5] = params.network;
  settings[6] = serial_port_rate + serial_parity_bit + air_data_rate;
  settings[7] = sub_packet_size + channel_noise + reserve + tx_power;
  settings[8] = params.channel;
  settings[9] = enable_rssi + transmission_mode + enable_repeater + enable_lbt + wor_control + wor_cycle;
  settings[10] = params.crypth;
  settings[11] = params.cryptl;

  char result[12];
  write_serial(settings, 12, result);

  Serial.println("result config : ");
  print_hex(result, 12);

  if (result[0] == 0xff && result[1] == 0xff && result[2] == 0xff) {
    Serial.println("[-] ERROR WRITING CONFIG");
  }

}

int sx126x::write_serial(char *data, int count, char *result)
{
  set_mode(MODE_CONF);

  if (_debug) {
    print_hex(data, count);
  }
  
  serial_lora.write(data, count);
  delay(100);

  count = receive(result);

  if (_debug) {
    print_hex(result, count);
    Serial.println();
  }

  set_mode(MODE_TRANS);
  return count;
}

void sx126x::print_hex(char* buffer, int count) {
  
  for (int i = 0; i < count; i++) {
    char x[3];
    sprintf(x, "%02x", buffer[i]);
    Serial.print(x);
    Serial.print(" ");
  }
  Serial.println();
}


s_rssi sx126x::get_rssi() {
  char data[] = {0xC0, 0xC1, 0xC2, 0xC3, 0x00, 0x02};
  
  //send using trans/WOR mode, NOT conf mode
  send(data, 6);
  char result[5];
  int count = receive(result);
  
  Serial.println("_______");
  Serial.println(count);
  print_hex(result, count);
  Serial.println("_______");

  struct s_rssi rssi;
  rssi.current_noise = result[3];
  rssi.rssi_last_receive = result[4];

  return rssi;
}



void sx126x::print_config() {

  Serial.print("\naddrh ");
  Serial.print(params.addrh);
  Serial.print("\naddrl ");
  Serial.print(params.addrl);
  Serial.print("\naddress ");
  Serial.print((params.addrh << 8) + params.addrl);
  Serial.print("\nnetwork ");
  Serial.print(params.network);
  
  Serial.print("\nserial_port_rate ");
  Serial.print(params.serial_port_rate);
  Serial.print("\nserial_parity_bit ");
  Serial.print(params.serial_parity_bit);

  Serial.print("\nair_data_rate ");
  Serial.print((float)params.air_data_rate / 10);
  Serial.print("\nsub_packet_size ");
  Serial.print(params.sub_packet_size);
  Serial.print("\nchannel_noise ");
  Serial.print(params.channel_noise);
  Serial.print("\ntx_power ");
  Serial.print(params.tx_power);
  Serial.print("\nchannel ");
  Serial.print(params.channel);

  Serial.print("\nenable_rssi ");
  Serial.print(params.enable_rssi);
  Serial.print("\ntransmission_mode ");
  Serial.print(params.transmission_mode ? "fixed" : "transparent");
  Serial.print("\nenable_repeater ");
  Serial.print(params.enable_repeater);
  Serial.print("\nenable_lbt ");
  Serial.print(params.enable_lbt);
  Serial.print("\nwor_control ");
  Serial.print(params.wor_control ? "transmitter" : "receiver");
  Serial.print("\nwor_cycle ");
  Serial.print(params.wor_cycle);

  Serial.print("\ncrypth ");
  Serial.print(params.crypth);
  Serial.print("\ncryptl ");
  Serial.print(params.cryptl);

  Serial.println();
}